dashboard/src/actions/kube.tsx

```diff
@@ -15,25 +15,45 @@ export const receiveResourceError = createAction("RECEIVE_RESOURCE_ERROR", resol
   return (resource: { key: string; error: Error }) => resolve(resource);
 });
 
-const allActions = [requestResource, receiveResource, receiveResourceError];
+export const openWatchResource = createAction("OPEN_WATCH_RESOURCE", resolve => {
+  return (ref: ResourceRef) => resolve(ref);
+});
+
+const allActions = [requestResource, receiveResource, receiveResourceError, openWatchResource];
 
 export type KubeAction = ActionType<typeof allActions[number]>;
 
 export function getResource(
-  apiVersion: string,
-  resource: string,
-  namespace: string,
-  name: string,
-  query?: string,
+  ref: ResourceRef,
 ): ThunkAction<Promise<void>, IStoreState, null, KubeAction> {
-  return async dispatch => {
-    const key = Kube.getResourceURL(apiVersion, resource, namespace, name, query);
+  return async (dispatch, getState) => {
+    const key = ref.getResourceURL();
+
+    // Multiple requests for a resource may happen at the same time (e.g. when
+    // loading different sections of a view). This guard attempts to prevent
+    // multiple requests for a resource that is already being fetched. Since
+    // this action is asynchronous, it's possible that multiple requests are let
+    // through, but this is not a huge concern.
+    const existing = getState().kube.items[key];
+    if (existing && existing.isFetching) {
```

Comment:
I think we should only check if `existing` is set. If multiple requests happens sequentially we don't want to repeat the request either, isn't it? If we want to get updates fro a resource it should use a socket?

Reply:
I considered it, but I think it's helpful to refresh the data on a new page/component load, and this is the same as what happens today. I think the main concern is not repeating requests around the same time, not preventing requests that we might want to happen at different times.

Reply:
Yes, I am thinking about requests around the same time. Depending on the performance both components (e.g. ServiceTable and AccessURLTable) may request a service with enough delay for the first request to be completely fetched before the second request. The problem in that case is that the second component will show the loading icon even if the resource has just been collected.

I think that the UX would be a bit better if the second time you visit an app the resources are already loaded. If there are updates the sockets will handle those changes eventually.

Reply:
Okay, I see what you're saying. I think that's fairly unlikely to happen, given how fast dispatching actions are compared to API requests. We should keep an eye on this though in case that does happen.

I understand your point about UX, but I don't think it's a big concern. Fetching the resources outside of the WebSockets has already improved the loading speed substantially. If we were to do this, we would need to either create another action of have a `skipExisting` parameter in this action as not all resources will open a WebSocket to get the latest data (e.g. Secrets).

Reply:
Just to clarify, you're okay leaving this as-is for now? I'm happy to revisit caching this if we see performance problems, but I would prefer we don't prematurely optimise here.

Reply:
Yes, I mean that you can go ahead with this code.